### PR TITLE
fix:[DEL-5969]: Failing all running task associated to a delegate if same delegate registration happens

### PR DESCRIPTION
**Pull request by Vbhatia20**
## Describe your changes
Goal: We want to fail/end all running task associated to a delegate if same delegate registration happens
Solution: 
1. During execution of registration, if we find delegate already exists, we make a call to mark all task as failed.
2. We are catching the exception here as we don't want user to be blocked even if failing of task is not successful for some reason.

Tests: (In Progress)
1. Install an Immutable k8 delegate and execute task and kill delegate process for it to come up again. Here new host came up.
2. Tested upgrade of delegate version and registration.
3. Tested upgrade and downgrade of legacy delegate during multiple task execution
Screenshot 2023-02-16 at 2 24 49 PM

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-ALL: `trigger tiAll`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/44023)

### CODE REVIEW

File: 400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java [hunk 1]
```diff
@@ -2601,6 +2601,16 @@ public DelegateRegisterResponse register(final DelegateParams delegateParams, fi
     final Delegate existingDelegate = getExistingDelegate(delegateParams.getAccountId(), delegateParams.getHostName(),
         delegateParams.isNg(), delegateParams.getDelegateType(), delegateParams.getIp());
 
+    // this code is to mark all the task in running as failed if same delegate registration
+    if (existingDelegate != null) {
+      try {
+        onDelegateDisconnected(delegateParams.getAccountId(), existingDelegate.getDelegateGroupId());
+      } catch (Exception e) {
+        e.printStackTrace();
```
Comment (abhinav-harness) [inline 400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java:2609]: log error in warn not like this

Comment (Vbhatia20) [reply 400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java:2609]: modified to error

**Review by mdmilic — COMMENTED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44023)* status: Some of the checks are too old. 1. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java` line 2605 at r2](https://reviewable.io/reviews/harness/harness-core/44023#-NOj5aGPCqo5lyoHSfx8:-NOj5aGPCqo5lyoHSfx9:b-9rymdp) ([raw file](https://github.com/harness/harness-core/blob/15b52e2a60696fd8de20705ee6ab99c3e3f19da7/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L2605)):*
> ```Java
> 
>     // this code is to mark all the task in running as failed if same delegate registration
>     if (existingDelegate != null) {
> ```

Have you tested this with legacy delegates upgrade scenarios? Do we want to do this only for immutable delegates or for all (or is it guaranteed that we are immutable at this point)? Legacy delegates are free to disconnect and re-connect re-register. e.g. This normally happens when legacy delegate upgrades, the new process that took over should be able to execute tasks, not sure if it can be affected. 

Additionally for immutable we saw weird behavior for immutable in this scenario when customer was abusing delegate groups in CG. Do we need to clear old records in this scenario (for immutable)?
___
*[`400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java` line 2610 at r2](https://reviewable.io/reviews/harness/harness-core/44023#-NOj6pYrC77wzEgP4zyG:-NOj6pYrC77wzEgP4zyH:b-1xzzbb) ([raw file](https://github.com/harness/harness-core/blob/15b52e2a60696fd8de20705ee6ab99c3e3f19da7/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L2610)):*
> ```Java
>       } catch (Exception e) {
>         e.printStackTrace();
>         log.error("couldn't delete the task associated with existing delegate: "+existingDelegate.getDelegateGroupId());
> ```

Remove `e.printStackTrace();` and add exception as a param to the error log (which will print stacktrace in logs). Also don't use string concatenation with logging, use variable placeholders. This is good for performance as it doesn't create extra String objects (it doesn't matter in this instance since this exception should rarely happen, but it's a good practice to do it in general)



_Suggestion:_
```Java
        log.error("Couldn't delete the task associated with existing delegate: {}", existingDelegate.getDelegateGroupId(), e);
```

**Review by Vbhatia20 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44023)* status: 1 file(s) to review. 1. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java` line 2609 at r1](https://reviewable.io/reviews/harness/harness-core/44023#-NOMHThPaeWu-f31FyxR-r1-2609:-NOl24ct1e4HcUEf7FCd:b-7rl0kd) ([raw file](https://github.com/harness/harness-core/blob/f702370c3aac1caeb0b32543e877c15e9795d91d/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L2609)):*
<details><summary><i>Previously, Vbhatia20 (Vishal Bhatia) wrote…</i></summary><blockquote>

modified to error
</blockquote></details>

done
___
*[`400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java` line 2605 at r2](https://reviewable.io/reviews/harness/harness-core/44023#-NOj5aGPCqo5lyoHSfx8:-NOl3qSmAQgKeLu06K0e:botsnme) ([raw file](https://github.com/harness/harness-core/blob/15b52e2a60696fd8de20705ee6ab99c3e3f19da7/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L2605)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

Have you tested this with legacy delegates upgrade scenarios? Do we want to do this only for immutable delegates or for all (or is it guaranteed that we are immutable at this point)? Legacy delegates are free to disconnect and re-connect re-register. e.g. This normally happens when legacy delegate upgrades, the new process that took over should be able to execute tasks, not sure if it can be affected. 

Additionally for immutable we saw weird behavior for immutable in this scenario when customer was abusing delegate groups in CG. Do we need to clear old records in this scenario (for immutable)?
</blockquote></details>

I have tested upgrade and downgrade of legacy delegate with multiple task execution happening during upgrade. The upgrade was done only post all the tasks were executed and new tasks started execution on new version only when it became healthy 
___
*[`400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java` line 2610 at r2](https://reviewable.io/reviews/harness/harness-core/44023#-NOj6pYrC77wzEgP4zyG:-NOl2AMt1_mPZtvNPx2_:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/15b52e2a60696fd8de20705ee6ab99c3e3f19da7/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L2610)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

Remove `e.printStackTrace();` and add exception as a param to the error log (which will print stacktrace in logs). Also don't use string concatenation with logging, use variable placeholders. This is good for performance as it doesn't create extra String objects (it doesn't matter in this instance since this exception should rarely happen, but it's a good practice to do it in general)
</blockquote></details>

done

**Review by mdmilic — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44023#-:-NOoLbMZCMDiYYhlFnDJ:bnfp4nl)* status: 1. unresolved discussions.

**Review by AnupamIO — APPROVED** [review]
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44023)* status: 1. unresolved discussions.

**Comment by Vbhatia20** [comment]
trigger recency.

**Comment by Vbhatia20** [comment]
trigger messagecheck

**Comment by xingchi-jin** [comment]
we had a discussion on the call. I will approve after we verify one scenario,
When upgrading legacy delegates, the running tasks are able to finish and not impacted by this change.

**Review by xingchi-jin — APPROVED** [review]
> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44023)* status: 1. unresolved discussions.

**Comment by xingchi-jin** [comment]
fix code format and let's merge it

**Review by xingchi-jin — APPROVED** [review]
> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44023)* status: 1. unresolved discussions.

File: 400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java [hunk 2]
```diff
@@ -2601,6 +2601,18 @@ public DelegateRegisterResponse register(final DelegateParams delegateParams, fi
     final Delegate existingDelegate = getExistingDelegate(delegateParams.getAccountId(), delegateParams.getHostName(),
         delegateParams.isNg(), delegateParams.getDelegateType(), delegateParams.getIp());
 
+    // this code is to mark all the task in running as failed if same delegate registration for immutable
+    // this should not impact any functionality wrt legacy delegate
+    if ((existingDelegate != null) && (existingDelegate.isImmutable())) {
+      try {
+        log.info("disconnecting task associated with existing delegate");
+        onDelegateDisconnected(delegateParams.getAccountId(), existingDelegate.getHostName());
```
Comment (xingchi-jin) [inline 400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java:2609]: use delegate.getUuid()

**Review by xingchi-jin — APPROVED** [review]
> Reviewed 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44023)* status: 1. unresolved discussions.

**Review by xingchi-jin — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44023)* status: 1. unresolved discussions.

**Pull request closed without merging**